src/library/scala/collection/mutable/Map.scala

```diff
@@ -199,8 +199,17 @@ trait MapOps[K, V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
     * @return   the map itself.
     */
   def mapValuesInPlace(f: (K, V) => V): this.type = {
-    iterator foreach {
-      case (key, value) => update(key, f(key, value))
+    if (nonEmpty) this match {
+      case hm: mutable.HashMap[K, V] => hm.mapValuesInPlaceImpl(f)
+      case _ =>
+        val array = this.toArray[Any]
```

Comment:
We should add some more overrides (in 2.14 or as special cases here in 2.13) to avoid this default implementation.

Reply:
should we add overrides to `filterInPlace` as well?